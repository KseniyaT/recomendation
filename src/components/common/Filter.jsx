import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Menu from '../common/Menu';

class Filter extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isMenuOpen: false,
    };
    this.handleFilterIconClick = this.handleFilterIconClick.bind(this);
  }

  /**
   * It shows/hide dropdown
   * @returns {void}
   */
  handleFilterIconClick() {
    this.setState({
      isMenuOpen: !this.state.isMenuOpen,
    });
  }

  render() {
    const { isMenuOpen } = this.state;
    const { list, isDisabled, chosenId, onClick, onScroll } = this.props;

    return (
      <div className="filter filter_left">
        <button
          className="filter__icon"
          disabled={isDisabled}
          onClick={this.handleFilterIconClick}
        >
          <i className="material-icons">filter_list</i>
        </button>
        {
          isMenuOpen &&
          <div className="filter__dropdown">
            <Menu
              list={list}
              chosenId={chosenId}
              onClick={onClick}
              onScroll={onScroll}
            />
          </div>
        }
      </div>
    );
  }
}

Filter.propTypes = {
  list: PropTypes.array,
  chosenId: PropTypes.any,
  isDisabled: PropTypes.bool,
  onClick: PropTypes.func,
  onScroll: PropTypes.func,
};

export default Filter;
