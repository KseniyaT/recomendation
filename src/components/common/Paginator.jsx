import React from 'react';
import PropTypes from 'prop-types';

function Paginator(props) { // eslint-disable-line require-jsdoc
  if (!props.pageNumber || +props.pageNumber < 0) return null;

  const handleItemClick = ({ event, index }) => {
    event.preventDefault();
    if (index === props.currentPage) return;
    typeof props.onClick === 'function' && props.onClick(index);
  };

  const paginationItems = [];
  for (let index = 0; index < props.pageNumber; index++) {
    paginationItems.push(
      <li className="paginator__element" key={index}>
        <a
          className={
            props.currentPage === index ? 'paginator__link paginator__link_active': 'paginator__link'
          }
          onClick={(event) => handleItemClick({ event, index })}
          type="button"
        >
          {index+1}
        </a>
      </li>
    );
  }

  return <ul className="paginator">{paginationItems}</ul>;
}

Paginator.propTypes = {
  pageNumber: PropTypes.number,
  currentPage: PropTypes.number,
  onClick: PropTypes.func,
};

export default Paginator;
