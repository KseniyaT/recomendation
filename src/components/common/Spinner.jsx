import React from 'react';

function Spinner() { // eslint-disable-line require-jsdoc
  return (
    <div className="spinner">
      <div className="spinner__inner">
        <div className="spinner__inner__heard" />
      </div>
    </div>
  );
}

export default Spinner;
