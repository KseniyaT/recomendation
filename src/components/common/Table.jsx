import React from 'react';
import PropTypes from 'prop-types';
import utils from '../../helpers/utils';

function Table(props) { // eslint-disable-line require-jsdoc
  const { theadList, tbodyList } = props;
  if (!(tbodyList && tbodyList.length)) return null;
  return (
    <table className="table">
      <thead className="table__head">
        <tr className="table__row">
          {
            theadList.map((th) => {
              return <th className="table__cell table__cell_th" key={th.id}>{th.label}</th>;
            })
          }
        </tr>
      </thead>
      <tbody className="table__body">
        {
          tbodyList.map((tbody) => {
            const content = tbody.content;
            return (
              <tr className="table__row" key={tbody.id}>
                {
                  content.map((td) => {
                    return (<td className="table__cell" key={utils.generateUniqKey(tbody.id)}>{td}</td>);
                  })
                }
              </tr>
            );
          })
        }
      </tbody>
    </table>
  );
}

Table.propTypes = {
  theadList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    label: PropTypes.any,
  })),
  tbodyList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    content: PropTypes.array,
  })),
};

export default Table;
