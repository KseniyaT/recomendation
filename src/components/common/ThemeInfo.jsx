import React from 'react';
import PropTypes from 'prop-types';

function ThemeInfo(props) { // eslint-disable-line require-jsdoc
  const { sentiment, label, isLoading } = props;
  let iconLigature = 'sentiment_very_satisfied';
  if (sentiment === 0) {
    iconLigature = 'sentiment_satisfied';
  } else if (sentiment === -1) {
    iconLigature = 'sentiment_very_dissatisfied';
  }
  return (
    <div className="theme-info">
      <p className="theme-info__name">{isLoading ? '...Loading' : label}</p>
      <i className="theme-info__icon material-icons">{iconLigature}</i>
    </div>
  );
}

ThemeInfo.propTypes = {
  sentiment: PropTypes.number,
  label: PropTypes.any,
  isLoading: PropTypes.bool,
};

export default ThemeInfo;
