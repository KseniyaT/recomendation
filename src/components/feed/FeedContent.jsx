import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Spinner from '../common/Spinner';
import Paginator from '../common/Paginator';
import Filter from '../common/Filter';
import Table from '../common/Table';
import ThemeInfo from '../common/ThemeInfo';
import reviewActions from '../../actions/reviews/index';
import themeActions from '../../actions/themes/index';
import CONSTANTS from '../../constants/index';
import utils from '../../helpers/utils';

const { getReviews } = reviewActions;
const { getThemes } = themeActions;

class Feed extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filterValue: '',
      reviewOffset: 0,
      maxPageNumber: 0,
      filterId: undefined,
      reviewLimit: CONSTANTS.DEFAULTS.REVIEW_LIMIT,
      themeLimit: CONSTANTS.DEFAULTS.THEME_LIMIT,
    };
    this.handleFilterClick = this.handleFilterClick.bind(this);
    this.handleLoadNextLinesClick = this.handleLoadNextLinesClick.bind(this);
    this.handleMenuScroll = this.handleMenuScroll.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
  }

  componentDidMount() {
    this.props.getThemes();
    this.props.getReviews();
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.reviewsList !== this.props.reviewsList) {
      const reviewOffset = prevState.reviewOffset + this.props.reviewsList.length;
      const newPageNumber = Math.floor(reviewOffset/prevState.reviewLimit);
      const maxPageNumber = prevState.maxPageNumber > newPageNumber ?
        prevState.maxPageNumber :
        newPageNumber;
      this.setState({
        reviewOffset,
        maxPageNumber,
      });
    }
  }

  /**
   * It calls reviews request with theme value params
   * when an user clicks on an element in filter list or on remove filter btn
   * @param {String} value - name of theme
   * @param {*} id - unique id of theme
   * @returns {void}
   */
  handleFilterClick({ value, id }) {
    let filterId = id;
    let filterValue = value;
    if (this.state.filterId === id) {
      filterId = undefined;
      filterValue = '';
    }
    this.setState({
      filterValue,
      filterId,
      reviewOffset: 0,
      maxPageNumber: 0,
      reviewLimit: CONSTANTS.DEFAULTS.REVIEW_LIMIT,
    });
    const params = id && this.state.filterId !== id ? { theme_id: id } : null;
    !this.props.isReviewsLoading && this.props.getReviews(params);
  }

  /**
   * It asks for next {revoewLimit} lines to show it
   * @returns {void}
   */
  handleLoadNextLinesClick() {
    !this.props.isReviewsLoading &&
    this.props.getReviews({
      offset: this.state.reviewOffset,
      limit: this.state.reviewLimit,
      ...(this.state.filterId && { theme_id: this.state.filterId }),
    });
  }

  /**
   * It asks for next {themeLimit} lines to show it in dropdown
   * when an user scrolls it and scrollTop is more than 10/13 of the dropdonw
   * @param {Number} height - height of elemets in dropdown
   * @param {Number} scrollTop - scroll position in dropdown
   * @returns {void}
   */
  handleMenuScroll({ height, scrollTop }) {
    if (scrollTop > height/1.3) {
      !this.props.isThemeListLoading &&
      this.props.getThemes({ offset: this.props.themesOffset, limit: this.state.themeLimit });
    }
  }

  handlePageClick(index) {
    if (!this.props.isReviewsLoading) {
      const offset = index * this.state.reviewLimit;
      this.setState({
        reviewOffset: offset,
      });
      !this.props.isReviewsLoading &&
      this.props.getReviews({
        offset: index * this.state.reviewLimit,
        limit: this.state.reviewLimit,
        ...(this.state.filterId && { theme_id: this.state.filterId }),
      });
    }
  }

  render() {
    const { filterValue, filterId, reviewLimit, reviewOffset, maxPageNumber } = this.state;
    const {
      specificThemeObj, reviewsList, themesList,
      isThemeListLoading, isReviewsLoading, isThemeObjLoading,
    } = this.props;

    const isLoading = !reviewsList.length && isReviewsLoading;
    const currentPage = Math.floor(reviewOffset / reviewLimit) - 1;

    const content = isLoading ?
      <Spinner /> :
      <div className="wrapper">
        <div className="wrapper__header">
          {
            filterValue &&
            <p className="wrapper__title">
              <span className="wrapper__title__text">Sort by: {filterValue}</span>
              <a
                className="wrapper__title__icon-container" type="button"
                onClick={() => this.handleFilterClick({ value: '', id: undefined })}
              >
                <i className="wrapper__title__icon material-icons">close</i>
              </a>
            </p>
          }
          <Filter
            list={themesList}
            chosenId={filterId}
            isDisabled={isThemeListLoading}
            onClick={this.handleFilterClick}
            onScroll={this.handleMenuScroll}
          />

        </div>
        <div className="wrapper__body">
          {
            reviewsList.length ?
              <Table
                theadList={[{ id: 0, label: 'Theme' }, { id: 1, label: 'Feedback' }]}
                tbodyList={
                  reviewsList.map((reviewObj) => {
                    return {
                      id: reviewObj.id,
                      content: [
                        reviewObj.themes && reviewObj.themes.map((theme) => {
                          return (
                            <ThemeInfo
                              key={utils.generateUniqKey(`${reviewObj.id}_${theme.theme_id}`)}
                              sentiment={theme.sentiment}
                              label={specificThemeObj[theme.theme_id]}
                              isLoading={isThemeObjLoading}
                            />
                          );
                        }),
                        reviewObj.comment,
                      ],
                    };
                  })
                }
              />:
              <div>{CONSTANTS.MESSAGES.NOTHING_FOUND}</div>
          }
        </div>
        <div className="wrapper__footer">
          <Paginator pageNumber={maxPageNumber} currentPage={currentPage} onClick={this.handlePageClick} />
          {
            reviewsList.length === CONSTANTS.DEFAULTS.REVIEW_LIMIT &&
            <a
              className={isLoading ? 'link link_disabled' : 'link'}
              onClick={this.handleLoadNextLinesClick}
            >
              Load next {reviewLimit} lines
            </a>
          }
        </div>
      </div>;
    return (
      <section className="main">
        {content}
      </section>
    );
  }
}

Feed.propTypes = {
  themesList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  })),
  reviewsList: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    comment: PropTypes.string,
    themes: PropTypes.array,
  })),
  themesOffset: PropTypes.number,
  specificThemeObj: PropTypes.object,
  isThemeObjLoading: PropTypes.bool,
  isThemeListLoading: PropTypes.bool,
  isReviewsLoading: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  themesList: state.themes.themesList,
  themesOffset: state.themes.themesOffset,
  isThemeObjLoading: state.themes.isThemeObjLoading,
  isThemeListLoading: state.themes.isThemeListLoading,
  specificThemeObj: state.themes.specificThemeObj,
  reviewsList: state.reviews.reviewsList,
  isReviewsLoading: state.reviews.isLoading,
});

const mapDispatchToProps = (dispatch) => ({
  getReviews: (obj) => dispatch(getReviews(obj)),
  getThemes: (obj) => dispatch(getThemes(obj)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Feed);
