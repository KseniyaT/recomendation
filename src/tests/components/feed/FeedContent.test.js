import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import FeedContent from '../../../components/feed/FeedContent';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import Spinner from '../../../components/common/Spinner';
import Paginator from '../../../components/common/Paginator';
import Filter from '../../../components/common/Filter';
import Table from '../../../components/common/Table';
import ThemeInfo from '../../../components/common/ThemeInfo';
import CONSTANTS from '../../../constants/index';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('FeedContent', () => {
  let store;
  let rendererComponent;
  let wrapper;
  let wrapperClass;

  const defaultStore = {
    themes: {
      themesList: [
        { id: 6334, name: 'Accounts' },
        { id: 6335, name: 'Transactions' }
      ],
      themesOffset: 0,
      specificThemeObj: { 6334: 'Accounts', 6335: 'Transactions' },
      isThemeObjLoading: false,
      isThemeListLoading: false,
    },
    reviews: {
      reviewsList: [
        { id: 59458292, comment: "exelent", themes: [{ theme_id: 6334, sentiment: 0 }] },
        { id: 59457786, comment: "Excellent keep it up", themes: [{ theme_id: 6335, sentiment: 1 }] }
      ],
      isLoading: false,
    },
  };

  beforeEach(() => {
    store = mockStore(JSON.parse(JSON.stringify(defaultStore)));
    rendererComponent = renderer.create(
      <Provider store={store}>
        <FeedContent />
      </Provider>
    );

    store.dispatch = jest.fn();

    wrapperClass = mount(shallow(<FeedContent store={store} />).get(0));
  });

  it('should render with given state from Redux store', () => {
    expect(rendererComponent.toJSON()).toMatchSnapshot();
  });

  it('should call dispatch method when the paginator is clicked', () => {
    const fakeEvent = { preventDefault: () => {} };
    rendererComponent.root.findByType(Paginator).props.onClick(fakeEvent);
    expect(store.dispatch).toHaveBeenCalled();
  });

  it('should render all rendererComponents if data is loaded', () => {
    expect(wrapperClass.length).toBe(1);
    expect(wrapperClass.find(Paginator).length).toBe(1);
    expect(wrapperClass.find(Filter).length).toBe(1);
    expect(wrapperClass.find(Table).length).toBe(1);
    expect(wrapperClass.find(ThemeInfo).length).toBe(2);
    expect(wrapperClass.find('.wrapper').length).toBe(1);
    expect(wrapperClass.find(Spinner).length).toBe(0);
  });

  it('should show spinner whole data is loading', () => {
    const loadingStore = JSON.parse(JSON.stringify(defaultStore));
    const storeLoading = mockStore({ ...loadingStore, reviews: { reviewsList: [], isLoading: true } });
    const wrapperLoading = mount( <Provider store={storeLoading}> <FeedContent /> </Provider> );
    expect(wrapperLoading.length).toBe(1);
    expect(wrapperLoading.find(Spinner).length).toBe(1);
    expect(wrapperLoading.find('.wrapper').length).toBe(0);
  });

  it('should show according message if no review results', () => {
    const loadingStore = JSON.parse(JSON.stringify(defaultStore));
    const storeNoReview = mockStore({ ...loadingStore, reviews: { reviewsList: [], isLoading: false } });
    const wrapperNoReview = mount( <Provider store={storeNoReview}> <FeedContent /> </Provider> );
    expect(wrapperNoReview.length).toBe(1);
    expect(wrapperNoReview.find(Spinner).length).toBe(0);
    expect(wrapperNoReview.find('.wrapper__body > div').text()).toEqual(CONSTANTS.MESSAGES.NOTHING_FOUND);
  });

  it('should calls dispatch function when an user clicks on "Load more" link', () => {
    const list = [];
    for(let i = 0; i < 20; i++) {
      list.push({ id: i, comment: "exelent", themes: [{ theme_id: i, sentiment: 0 }] });
    }
    const longStore = JSON.parse(JSON.stringify(defaultStore));
    const longReview = mockStore({ ...longStore, reviews: { reviewsList: list } });
    const longWrapper = mount( <Provider store={longReview}> <FeedContent /> </Provider> );
    const a = longWrapper.find('.wrapper__footer').find('.link');
    a.simulate('click');
    expect(store.dispatch).toHaveBeenCalled();
  });

  it('should change filterValue state and call dispatch function when an user click on remove filter icon', () => {
    wrapperClass.setState({ filterValue: 'testFilter' });
    expect(wrapperClass.state('filterValue')).toEqual('testFilter');
    const a = wrapperClass.find('.wrapper__title > a');
    a.simulate('click');
    expect(wrapperClass.state('filterValue')).toEqual('');
    expect(store.dispatch).toHaveBeenCalled();
  });

  it('should call dispatch function when an user clicks on Filter (or scroll)', () => {
    wrapperClass.setState({ filterValue: 'testFilter2' });
    expect(wrapperClass.state('filterValue')).toEqual('testFilter2');

    wrapperClass.find(Filter).setState({ isMenuOpen: true });

    wrapperClass.find('.menu').find('.menu__link').first().simulate('click');
    expect(wrapperClass.state('filterValue')).toEqual('Accounts');
    expect(store.dispatch).toHaveBeenCalled();

    const mockedEvent = { target: { children: [{ clientHeight: 30 }], clientHeight: 20, scrollTop: 30 } };
    wrapperClass.find('.menu').simulate('scroll', mockedEvent);
    expect(store.dispatch).toHaveBeenCalled();
  });

});