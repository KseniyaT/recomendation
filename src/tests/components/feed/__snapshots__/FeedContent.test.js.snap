// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedContent should render with given state from Redux store 1`] = `
<section
  className="main"
>
  <div
    className="wrapper"
  >
    <div
      className="wrapper__header"
    >
      
      <div
        className="filter filter_left"
      >
        <button
          className="filter__icon"
          disabled={false}
          onClick={[Function]}
        >
          <i
            className="material-icons"
          >
            filter_list
          </i>
        </button>
      </div>
    </div>
    <div
      className="wrapper__body"
    >
      <table
        className="table"
      >
        <thead
          className="table__head"
        >
          <tr
            className="table__row"
          >
            <th
              className="table__cell table__cell_th"
            >
              Theme
            </th>
            <th
              className="table__cell table__cell_th"
            >
              Feedback
            </th>
          </tr>
        </thead>
        <tbody
          className="table__body"
        >
          <tr
            className="table__row"
          >
            <td
              className="table__cell"
            >
              <div
                className="theme-info"
              >
                <p
                  className="theme-info__name"
                >
                  Accounts
                </p>
                <i
                  className="theme-info__icon material-icons"
                >
                  sentiment_satisfied
                </i>
              </div>
            </td>
            <td
              className="table__cell"
            >
              exelent
            </td>
          </tr>
          <tr
            className="table__row"
          >
            <td
              className="table__cell"
            >
              <div
                className="theme-info"
              >
                <p
                  className="theme-info__name"
                >
                  Transactions
                </p>
                <i
                  className="theme-info__icon material-icons"
                >
                  sentiment_very_satisfied
                </i>
              </div>
            </td>
            <td
              className="table__cell"
            >
              Excellent keep it up
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="wrapper__footer"
    />
  </div>
</section>
`;
